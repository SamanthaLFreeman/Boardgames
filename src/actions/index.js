export const getGames = games => ({
  type: 'GET_GAMES',
  games
})

export const getCategories = categories => ({
  type: 'GET_CATEGORIES',
  categories
})

export const addFavorite = id => ({
  type: 'ADD_FAVORITE',
  id
})

export const removeFavorite = id => ({
  type: 'REMOVE_FAVORITE',
  id
})