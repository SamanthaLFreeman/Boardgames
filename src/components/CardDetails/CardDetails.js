import React from 'react';
import { connect } from 'react-redux';

export const CardDetails = ({ name, year_published, min_players, max_players, max_playtime, description_preview, image_url, primary_publisher, categoriesIds, rules_url, categoriesKey}) => {
  const displayCategories = categoriesIds.map(category => category.id).reduce((acc, id) => {
    categoriesKey.forEach(key => {
      if(id === key.id) {
        acc.push(key.name)
      }
    })
    return acc;
  }, []).map(categoryName => <li key={categoryName}>{categoryName}</li>)
  return (
    <section>
      <img src={image_url} alt={name} />
      <h3>{name}</h3>
      <p>Year Published: {year_published}</p>
      <p>Min Players: {min_players}</p>
      <p>Max Players: {max_players}</p>
      <p>Max Playtime: {max_playtime}</p>
      <p>{description_preview}</p>
      <p>Primary Publisher: {primary_publisher}</p>
      <p>Categories:</p>
      <ul>
        {displayCategories}
      </ul>
      <p>Rules: {rules_url}</p>
    </section>
  )
}

export const mapStateToProps = state => ({
  categoriesKey: state.categories
})

export default connect(mapStateToProps, null)(CardDetails);