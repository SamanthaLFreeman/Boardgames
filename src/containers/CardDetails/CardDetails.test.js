import React from 'react';
import { shallow } from 'enzyme';
import { CardDetails, mapStateToProps, mapDispatchToProps } from './CardDetails';
import { updateGame } from '../../actions';

describe('CardDetails', () => {
  let wrapper;
  const mockCard = {
    name: 'Takenoko',
    id: 41
  }
  const categoriesIdsMock = [
    {
      id: 1234
    }
  ]
  const categoriesKeyMock = [
    {
      id: 1234,
      name: "Fantasy"
    }
  ]
  const ownedGamesMock = [
    {
      name: 'Takenoko',
      id: 41
    }
  ]
  const mockState = {
    games: [mockCard],
    categories: categoriesKeyMock,
    ownedGames: ownedGamesMock,
    favorites: [mockCard]
  }
  const updateGameMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<CardDetails 
    card={mockCard}
    id={41}
    categoriesIds={categoriesIdsMock}
    categoriesKey={categoriesKeyMock}
    ownedGames={ownedGamesMock}
    updateGame={updateGameMock} />)
  });

  it('should match the snapshot with the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when handleChange is called', () => {
    const mockEvent = { target: { name: 'house_rules', value: 'Start with 7 cards' } };
    const expected = 'Start with 7 cards';

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('house_rules')).toEqual(expected);
  });

  it('should run handleSubmit when the button is clicked', () => {
    wrapper.instance().handleSubmit = jest.fn();
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.instance().forceUpdate();
    wrapper.find('button').at(1).simulate('click', mockEvent);

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });

  it('mapStateToProps should grab the props it needs', () => {
    const expected = {
      categoriesKey: categoriesKeyMock,
      ownedGames: ownedGamesMock
    }

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });

  it('it calls dispatch with the updateGame action', () => {
    const updatedGameMock = {
      name: 'Takenoko',
      id: 42,
      house_rules: 'Name the Panda'
    }
    const mockDispatch = jest.fn();
    const actionToDispatch = updateGame(updatedGameMock);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.updateGame(updatedGameMock);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
})