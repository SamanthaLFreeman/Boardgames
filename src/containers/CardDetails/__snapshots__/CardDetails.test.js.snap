// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetails should match the snapshot with the data passed through 1`] = `
<section
  className="card-details"
>
  <Link
    className="back-btn"
    to="/"
  >
    <button
      className="back-btn"
    >
      Return to Games
    </button>
  </Link>
  <form
    className="edit-card"
  >
    <input
      className="edit-input"
      name="house_rules"
      onChange={[Function]}
      placeholder="Add house rules"
      type="text"
      value=""
    />
    <Link
      to="/card/41"
    >
      <button
        className="edit-btn"
        onClick={[Function]}
      >
        Add Changes
      </button>
    </Link>
  </form>
  <p
    className="card-descrip"
  >
    Categories:
  </p>
  <ul>
    <li
      className="card-info"
      key="Fantasy"
    >
      Fantasy
    </li>
  </ul>
</section>
`;
